pkg/syncer/syncers/downsyncer.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncers
+
+import (
+	"errors"
+	"fmt"
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/syncer/apis/edge/v1alpha1"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+)
+
+type DownSyncer struct {
+	logger                  klog.Logger
+	upstreamClientFactory   ClientFactory
+	downstreamClientFactory ClientFactory
+	upstreamClients         map[schema.GroupKind]*Client
+	downstreamClients       map[schema.GroupKind]*Client
+}
+
+func NewDownSyncer(logger klog.Logger, upstreamClientFactory ClientFactory, downstreamClientFactory ClientFactory, syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) (*DownSyncer, error) {
+
+	downSyncer := DownSyncer{
+		logger:                  logger,
+		upstreamClientFactory:   upstreamClientFactory,
+		downstreamClientFactory: downstreamClientFactory,
+	}
+
+	if err := downSyncer.initializeClients(syncedResources, conversions); err != nil {
+		logger.Error(err, "failed to initialize clients")
+		return &downSyncer, err
+	}
+
+	return &downSyncer, nil
+}
+
+func (ds *DownSyncer) initializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.upstreamClients = map[schema.GroupKind]*Client{}
+	ds.downstreamClients = map[schema.GroupKind]*Client{}
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) ReInitializeClients(syncedResources []edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	return initializeClients(ds.logger, syncedResources, ds.upstreamClientFactory, ds.downstreamClientFactory, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) getClients(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) (*Client, *Client, error) {
+	return getClients(resource, ds.upstreamClients, ds.downstreamClients, conversions)
+}
+
+func (ds *DownSyncer) SyncOne(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.logger.V(3).Info(fmt.Sprintf("downsync '%s'", resourceToString(resource)))
```

Comment:
If you use `%q` instead of `%s` then you do not have to supply the quotes in the source and you also get the usual escaping for special characters.